### PR TITLE
Fix Stream.mm to record in QT Player (#11)

### PR DESCRIPTION
**Pull request by mizt**
https://github.com/johnboiles/coremediaio-dal-minimal-example/pull/11

### CODE REVIEW

File: CMIOMinimalSample.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -43,6 +44,7 @@
 		29F2123624406AF200805A6D /* CoreVideo.framework */ = {isa = PBXFileReference; lastKnownFileType = wrapper.framework; name = CoreVideo.framework; path = System/Library/Frameworks/CoreVideo.framework; sourceTree = SDKROOT; };
 		29F2123824406AF800805A6D /* CoreGraphics.framework */ = {isa = PBXFileReference; lastKnownFileType = wrapper.framework; name = CoreGraphics.framework; path = System/Library/Frameworks/CoreGraphics.framework; sourceTree = SDKROOT; };
 		29F2123A24406B0300805A6D /* AppKit.framework */ = {isa = PBXFileReference; lastKnownFileType = wrapper.framework; name = AppKit.framework; path = System/Library/Frameworks/AppKit.framework; sourceTree = SDKROOT; };
+		E0CCED8C2464EFAE00D8AAF6 /* QuartzCore.framework */ = {isa = PBXFileReference; lastKnownFileType = wrapper.framework; name = QuartzCore.framework; path = System/Library/Frameworks/QuartzCore.framework; sourceTree = SDKROOT; };
```
Comment (johnboiles) [inline CMIOMinimalSample.xcodeproj/project.pbxproj:47]: Why is this necessary? I don't see any new `#import`s in the code

File: CMIOMinimalSample/Stream.mm [hunk 2]
```diff
@@ -171,22 +171,16 @@ - (void)fillFrame {
...
-    // timestamps and scales like mach_absolute_time() and NSEC_PER_SEC will work for display, but will error out
-    // when trying to record.
-    //
-    // Instead, we start our presentation times from zero (using the sequence number as a base), and use a scale that's
-    // a multiple of our framerate. This has been observed in parts of AVFoundation and lets us be frame-accurate even
-    // on non-round framerates (i.e., we can use a scale of 2997 for 29,97 fps content if we want to).
-    CMTimeScale scale = FPS * 100;
-    CMTime frameDuration = CMTimeMake(scale / FPS, scale);
-    CMTime pts = CMTimeMake(frameDuration.value * self.sequenceNumber, scale);
+    double hostTime = CACurrentMediaTime();
```
Comment (johnboiles) [inline CMIOMinimalSample/Stream.mm:174]: What's the difference in `CACurrentMediaTime() * NSEC_PER_SEC` and `mach_absolute_time()`? Shouldn't they be the same?

File: CMIOMinimalSample/Stream.mm [hunk 3]
```diff
@@ -171,22 +171,16 @@ - (void)fillFrame {
...
+    double timescale = 600.0;
+
+    CMTime eventTime = CMTimeMake(hostTime*timescale,timescale);
```
Comment (johnboiles) [inline CMIOMinimalSample/Stream.mm:177]: I think @iKenndac's solution in #12 is little better (using `frameDuration.value * self.sequenceNumber` for the timestamp) since it won't be affected by the timing when the system is able to execute this code. The timestamp shall be the timing when the frame _should_ be presented, not the timing when the OS happened to get back to this thread.

File: CMIOMinimalSample/Stream.mm [hunk 4]
```diff
@@ -171,22 +171,16 @@ - (void)fillFrame {
...
-    // when trying to record.
-    //
-    // Instead, we start our presentation times from zero (using the sequence number as a base), and use a scale that's
-    // a multiple of our framerate. This has been observed in parts of AVFoundation and lets us be frame-accurate even
-    // on non-round framerates (i.e., we can use a scale of 2997 for 29,97 fps content if we want to).
-    CMTimeScale scale = FPS * 100;
-    CMTime frameDuration = CMTimeMake(scale / FPS, scale);
-    CMTime pts = CMTimeMake(frameDuration.value * self.sequenceNumber, scale);
+    double hostTime = CACurrentMediaTime();
+    double timescale = 600.0;
```
Comment (johnboiles) [inline CMIOMinimalSample/Stream.mm:175]: Why is 600 better than 100?

**Comment by mizt** [comment]
The version that calculates frames by sequence number has a problem with audio input in QT Player.
At least this PR is working fine.
And recording also succeeds.

QuartzCore.framework is required for CACurrentMediaTime().

File: CMIOMinimalSample/Stream.mm (hunk 4 above)
Comment (mizt) [reply CMIOMinimalSample/Stream.mm:175]: Looking at various samples, the timescale value is often set to 600.

File: CMIOMinimalSample/Stream.mm (hunk 2 above)
Comment (mizt) [reply CMIOMinimalSample/Stream.mm:174]: The accuracy is different.

Comment (johnboiles) [reply CMIOMinimalSample/Stream.mm:174]: Why do you think that? From the Apple [docs](https://developer.apple.com/documentation/quartzcore/1395996-cacurrentmediatime?language=objc):

> A CFTimeInterval derived by calling mach_absolute_time() and converting the result to seconds.

**Comment by johnboiles** [comment]
So you're saying recording with audio wasn't working before your patch? What was happening? Was the audio not being included in the recording? Or was it not playing back correctly?

I could imagine this being possible if the `presentationTimeStamp` needs to line up with the timestamp of captured audio.

**Comment by mizt** [comment]
When using a built-in microphone（I'm singing).


327cfb58fc160ae063199796f142c215ffb5a264
https://github.com/johnboiles/coremediaio-dal-minimal-example/tree/327cfb58fc160ae063199796f142c215ffb5a264
Looks good (But the recording fails). 


23994ef7fc93e7e29a6c9d53b72020bd27954f38
https://github.com/johnboiles/coremediaio-dal-minimal-example/tree/23994ef7fc93e7e29a6c9d53b72020bd27954f38
Sound is not valid.
The animation will stop.

File: CMIOMinimalSample/Stream.mm (hunk 2 above)
Comment (mizt) [reply CMIOMinimalSample/Stream.mm:174]: `#import <Foundation/Foundation.h>
#import <mach/mach_time.h>
#import <QuartzCore/CoreAnimation.h>

int main(int argc, char *argv[]) {
	@autoreleasepool {
		NSLog(@"%llu",(uint64_t)(CACurrentMediaTime()*NSEC_PER_SEC));
		NSLog(@"%llu",mach_absolute_time());
	}
}`

2020-05-09 05:56:12.165 Timer[11518:763703] 46073695407692
2020-05-09 05:56:12.166 Timer[11518:763703] 46073696399730

Comment (mizt) [reply CMIOMinimalSample/Stream.mm:174]: It was my mistake.
The following code also worked correctly.
```
uint64_t hostTime = mach_absolute_time();
double timescale = 600.0;

CMTime eventTime = CMTimeMake((hostTime/(double)NSEC_PER_SEC)*timescale,timescale);
CMSampleTimingInfo timing;
timing.duration = CMTimeMake(timescale,FPS*timescale);
timing.presentationTimeStamp = eventTime;
timing.decodeTimeStamp = eventTime;
	   
OSStatus err = CMIOStreamClockPostTimingEvent(eventTime,hostTime,true,self.clock);
```
It was recorded correctly in QT Player.

File: CMIOMinimalSample/Stream.mm (hunk 4 above)
Comment (iKenndac) [reply CMIOMinimalSample/Stream.mm:175]: Those samples are likely running at 60fps, so a 600 result would be fps * 100, just like we're already using.
